File: Week_02/G20190343020279/LeetCode_94_0279.py
"""
 @method "颜色标记法"
 @explain 兼具栈迭代方法的高效，又像递归方法一样简洁易懂，更重要的是，这种方法对于前序、中序、后序遍历，能够写出完全一致的代码。1.使用颜色标记节点的状态，新节点为白色，已访问的节点为灰色。2.如果遇到的节点为白色，则将其标记为灰色，然后将其右子节点、自身、左子节点依次入栈。3.如果遇到的节点为灰色，则将节点的值输出。
 @complexity 时间复杂度：O(n) ,空间复杂度：O(n)
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        # 1.设定白灰两色
        WHITE,GRAY = 0,1
        # 2.用于返回结果的数组
        res = []
        # 3.维护一个带颜色属性的节点栈
        stack = [(WHITE,root)]
        while stack:
            # 4.每次遍历都会出栈一个节点
            color,node = stack.pop()
            if node is None:continue
            # 5.重点：根据栈先入后出，所以是右-根-左反过来写
            if color == WHITE:
                stack.append((WHITE,node.right))
                stack.append((GRAY,node))
                stack.append((WHITE,node.left))
            # 6.遇到灰节点输出
            else:
                res.append(node.val)
        return res

File: Week_02/G20190343020279/LeetCode_94_0279.js
/**
 * @method "栈遍历法"
 * @explain 基于维护一个栈来实现树的遍历，中序遍历相当于先把所有的最左边的节点都添加到栈里，当没有了之后就弹出最上面的节点，把这个节点的值加入到返回的数组，然后接着判断这个节点有没有右节点，没有就继续返回上一个节点直到遍历所有的节点。
 * @complexity 时间复杂度：O(n),空间复杂度：O(n)
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    // 1.用于返回结果的数组
    let res = []
    // 2.用于暂存节点的栈
    let stack = []
    // 3.用于放入当前节点
    let curr = root
    // 4.如果当前节点为空同时栈也为空则退出这个循环
    while (curr != null || stack.length != 0) {
        while (curr != null) {
            // 5.如果当前节点不为空，则先放到栈中备用
            stack.push(curr)
            // 6.拿出左节点继续遍历
            curr = curr.left
        }
        // 7.如果当前节点为空则拿出栈顶的节点
        curr = stack.pop()
        // 8.拿出后第一时间加入到返回结果的数组中
        res.push(curr.val)
        // 9.然后看看右节点
        curr = curr.right
    } 
    return res
};
